Allow storing points/coordinates as struct arrays in addition to interleaved layout (#26)
geoarrow/geoarrow: format.md

GeoArrow proposes a packed columnar data format for the fundamental geometry
types, using packed coordinate and offset arrays to define geometry objects.

The inner level is always a Struct array storing the
coordinate values (`x: [x, x, x, x], y: [y, y, y, y]`, physically stored as
one array per dimension with the name of each child array corresponding to
the dimension it represents (i.e., x, y, z, or m). For any geometry type except
Point, this inner level is nested in one or multiple
[variable sized list](https://arrow.apache.org/docs/format/Columnar.html#variable-size-list-layout)
arrays. In practice, this means we have additional arrays storing the offsets
that denote where a new geometry, a new geometry part, or a new polygon ring
starts.

> **Point**: `Struct<x: double, y: double, [z: double, [m: double>]]`

Reviewer: In this case, is it necessary for m to be a `double` at all? For example, if I have a Line representing the path of a ship on the ocean, I might want want me to be a `Date64`.

Reviewer: Or more ambitiously but still grounded in the actual example I have in mind, I might want m to be `struct<time: Date64, windspeed: float64, winddirection: dictionary, notes: string>`. Or I guess there's `t` as well. I might also just be asking for arbitrary keys in the array where only `x`, `y`, and `z` (?or not z?) are strictly typed.

GeoArrow proposes a packed columnar data format for the fundamental geometry
types, using packed coordinate and offset arrays to define geometry objects.

> The inner level is always an array of points. For any geometry type except

Reviewer 2: Total nit: I like having a distinction between "point" and "coordinate", where "point" is a geometry type and "coordinate" is just a minimal storage container. For example, the GeoJSON spec first defines ["position"](https://www.rfc-editor.org/rfc/rfc7946#section-3.1.1) and then separately [defines points in terms of positions](https://www.rfc-editor.org/rfc/rfc7946#section-3.1.2).

(This is also why I [didn't originally understand](https://github.com/geopandas/geo-arrow-spec/issues/20#issuecomment-1125311940) that https://github.com/geopandas/geo-arrow-spec/issues/16 referred to all geometry types, not just points)

GeoArrow proposes a packed columnar data format for the fundamental geometry
types, using packed coordinate and offset arrays to define geometry objects.

The inner level is always an array of coordinates. For any geometry type except
Point, this inner level is nested in one or multiple
[variable sized list](https://arrow.apache.org/docs/format/Columnar.html#variable-size-list-layout)
arrays. In practice, this means we have additional arrays storing the offsets
that denote where a new geometry, a new geometry part, or a new polygon ring
starts.

This specification supports coordinates encoded as a Struct array storing the
coordinate values as separate arrays (i.e., `x: [x, x, ...], y: [y, y, y, ...]`)
and a FixedSizeList of interleaved values (i.e., `[x, y, x, y, ...]`). As
implementations evolve, this specification may grow to support other coordinate
representations or shrink to support only one if supporting multiple
representations becomes a barrier to adoption.

> **Coordinate (Struct)**: `Struct<x: double, y: double, [z: double, [m: double>]]`

Reviewer 3: ```suggestion
**Coordinate (separated)**: `Struct<x: double, y: double, [z: double, [m: double>]]`
```

Do we want to use the "separated" term when talking about the non-interleaved layout for the coordinates? 
(a bit of a nitpick, but it's good to have clear and consistent terminology we can use throughout)

While "struct" is the name of the type in Arrow, it has the disadvantage that it can be slightly ambiguous in general (some might consider the interleaved layout actually an "array of structs", in context of the SoA vs AoS terms). And we still clearly list this layout to use a Struct type on the remainder of the line.

We already use "separated" in the paragraph above for the explanation of the struct-based layout, and I saw that @kylebarron is also using this term in the geoarrow-rs implementation (https://github.com/kylebarron/geoarrow-rs/blob/032f6a7099e9a8a3f2c1d0fe84a178e886e3e5b9/src/array/coord/combined/array.rs#L8-L24)

An array of coordinates can be stored as a Struct array containing two or more
child double arrays with names corresponding to the dimension represented by
the child. The first and second child arrays must represent the x and y
dimension; where z and m dimensions are both included, the z dimension must
preceed the m dimension.

> **Coordinate (FixedSizeList)**: `FixedSizeList<double>[n_dim]`

Reviewer 3: ```suggestion
**Coordinate (interleaved)**: `FixedSizeList<double>[n_dim]`
```

? (depending on what we want this term between brackets to mean, but in general I think it will be easier to talk about "interleaved coordinates" compared to "fixed-size list coordinates")

> An array of point geometries may also be represented by a single array

Reviewer 3: ```suggestion
An array of coordinates may also be represented by a single array
```

the child name of the middle list should be "rings"; the child name of the
inner list should be "vertices".

> **Well-known-binary (WKB)**: `Binary` or `LargeBinary` or `FixedSizeBinary`

Reviewer 3: Let's leave the addition of WKB (or WKT, or geojson?) for another PR, to focus here on the struct representation?

Truly mixed geometry types can be supported as a union of the other geometry
types, and it is planned to add a description of this later.

## Concrete examples of the memory layout

**Point (Struct)**

> Arrow type: `Struct<x: double, y: double, [z: double, [m: double>]]`

Reviewer 3: ```suggestion
Arrow type: `Struct<x: double, y: double>`
```

(for the specific example here, it's just xy)

```
[

>     [[{x: 0, y: 0}, {x: 0, y: 1}, {x: 0, y: 2}]],

Reviewer 3: @paleolimbot would you be OK with keeping the examples as is here? I personally find the current version a bit easier to read, and the example is mostly to show the additional nesting, regardless of the exact inner representation. We could add a note that for the struct version, all buffers are the same, except the coordinates buffer would be two buffers for x and y.
